source/opt/cfg_cleanup_pass.cpp

```diff
@@ -0,0 +1,186 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a pass to cleanup the CFG to remove superfluous
+// constructs (e.g., unreachable basic blocks, empty control flow structures,
+// etc)
+
+#include <queue>
+#include <unordered_set>
+
+#include "cfg_cleanup_pass.h"
+
+#include "function.h"
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// Remove Phi operand {|rm_arg_id|, |rm_block_id|} from |phi|. This works by
+// building a new operand vector for |phi|.  The new operand vector will have
+// all the original operands except the operand {|rm_arg_id|, |rm_block_id|}.
+void RemovePhiOperand(ir::Instruction* phi, uint32_t rm_arg_id,
+                      uint32_t rm_block_id) {
+  std::vector<ir::Operand> keep_operands;
+
+  // Traverse all the operands in |phi|. Build the new operand vector by adding
+  // all the original operands from |phi| except the unwanted one.
+  for (uint32_t i = 0; i < phi->NumOperands();) {
+    if (i < 2) {
+      // The first two arguments are always preserved.
+      keep_operands.push_back(phi->GetOperand(i));
+      ++i;
+    } else {
+      // The remaining Phi arguments come in pairs. Index 'i' contains the
+      // variable id, index 'i + 1' is the originating block id. If these two
+      // match {|rm_arg_id|, |rm_block_id|}, skip them.
+      assert(i % 2 == 0 && i < phi->NumOperands() - 1 &&
+             "malformed Phi arguments");
+      if (rm_arg_id != phi->GetSingleWordOperand(i) ||
+          rm_block_id != phi->GetSingleWordOperand(i + 1)) {
+        keep_operands.push_back(phi->GetOperand(i));
+        keep_operands.push_back(phi->GetOperand(i + 1));
+      }
+
+      i += 2;
+    }
+  }
+
+  phi->ReplaceOperands(keep_operands);
+}
+
+}  // namespace
+
+void CFGCleanupPass::RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
+                                                  ir::Instruction* inst) {
+  uint32_t inst_id = inst->result_id();
+  if (inst_id == 0) {
+    return;
+  }
+
+  analysis::UseList* uses = def_use_mgr_->GetUses(inst_id);
+  if (uses == nullptr) {
+    return;
+  }
+
+  for (auto u : *uses) {
+    if (u.inst->opcode() == SpvOpPhi) {
+      RemovePhiOperand(u.inst, inst_id, block.id());
+    }
+  }
+}
+
+void CFGCleanupPass::RemoveBlock(ir::Function::iterator* bi) {
+  auto& block = **bi;
+  block.ForEachInst([&block, this](ir::Instruction* inst) {
+    // Note that we do not kill the block label instruction here. The label
+    // instruction is needed to identify the block, which is needed by the
+    // removal of PHI operands.
+    if (inst != block.GetLabelInst()) {
+      RemoveFromReachedPhiOperands(block, inst);
+      KillNamesAndDecorates(inst);
+      def_use_mgr_->KillInst(inst);
+    }
+  });
+
+  // Remove the label instruction last.
+  auto label = block.GetLabelInst();
+  KillNamesAndDecorates(label);
+  def_use_mgr_->KillInst(label);
```

Comment:
You'll also need to remove this label from any phis that might use it.
E.g. if removing block A, then remove any pairs (X A) from any Phis.   For example X could be a constant value (as in the example I gave earlier.)

Reply:
Still need to take care of this.

If RemovePhiOperand allowed a 0 arg id then you could use that.